### ISSUE TITLE
Make `file` source async

### ISSUE THREAD

**Issue by jszwedko** [vectordotdev/vector#9407]
The `file` source should implement `AsyncRead` so that we can integrate `FramedRead` and `codecs::Decoder`.

**Comment by tamer-hassan** [comment]
If it matters: in https://github.com/vectordotdev/vector/pull/22803 -- enhancement(file source): Use notify-rs/notify for filesystem notifications, `file` source now uses async/await for better performance and resource utilization

**Comment by tamer-hassan** [comment]
> If it matters: in [#22803](https://github.com/vectordotdev/vector/pull/22803) -- feat(sources): Add new ifile source with improved implementation: `ifile` source now uses async/await for better performance and resource utilization

While implementing `AsyncRead` is technically possible, I'd like to avoid it for these reasons:

1. The current architecture is specifically designed for file watching and checkpointing, which would be harder to maintain with a generic streaming interface
2. Besides the challenges of maintaining checkpoint functionality and preserving existing features like compression detection/handling and multi-line aggregation, the existing implementation also handles important edge cases like:
- Partial reads
- File rotations
- Dead file detection
- Custom line delimiting
- Position tracking
3. The benefits of using FramedRead and codecs would be outweighed by the complexity of maintaining Vector's specific requirements within the AsyncRead constraints

To summarize: I believe that the current implementation, while more specialized, better serves Vector's specific use case of reliable file monitoring and ingestion.

@pront any thoughts on this?
